docs/logging.md

```diff
@@ -0,0 +1,93 @@
+# Logging
+
+By default LinuxKit will write onboot and service logs directly to files in
+`/var/log` and `/var/log/onboot`.
+
+It is tricky to write the logs to a disk or a network service as no disks
+or networks are available until the `onboot` containers run. We work around
+this by splitting the logging into 2 pieces:
+
+1. `memlogd`: an in-memory circular buffer which receives logs (including
+   all the early `onboot` logs)
+2. a log writing `service` that starts later and can download and process
+   the logs from `memlogd`
+
+To use this new logging system, you should add the `memlogd` container to
+the `init` block in the LinuxKit yml. On boot `memlogd` will be started
+from `init.d` and it will listen on a Unix domain socket:
+
+```
+/var/run/external-logging.sock
+```
+
+The `init`/`service` process will look for this socket and redirect the
+`stdout` and `stderr` of both `onboot` and `services` to `memlogd`.
+
+## memlogd: an in-memory circular buffer
+
+The `memlogd` daemon reads the logs from the `onboot` and `services` containers
+and stores them together with a timestamp and the name of the originating
+container in a circular buffer in memory.
+
+The contents of the circular buffer can be read over the Unix domain socket
+```
+/var/run/memlogq.sock
+```
+
+The circular buffer has a fixed size (overridden by the command-line argument
+`-max-lines`) and when it fills up, the oldest messages will be overwritten.
+
+To store the logs somewhere more permanent, for example a disk or a remote
+network service, a service should be added to the yaml which connects to
+`memlogd` and streams the logs. The example program `logread` in the `memlogd`
+package demonstrates how to do this.
+
+### Message format
+
+The format used to read logs is similar to [kmsg](https://www.kernel.org/doc/Documentation/ABI/testing/dev-kmsg):
+```
+<timestamp>,<origin>;<body>
+```
+where `<timestamp>` is an RFC3339-formatted timestamp, `<origin>` is the name of the
+original sending process and `<body>` is the output.
```

Comment:
What is `<origin>` is it the processes `argv[0]` at time of launch, `argv[0]` now or something else? In principal it could contain a `;` which would royally confuse the recipient. Might the `pid` be sufficient?

The examples below suggest it might actually be the name of the log file rather than the process, in which case just the text needs changing, assuming we agree that `;` is verbotten in a service name...

Reply:
It's actually the name of the log -- it's chosen by `pkg/init` when it sets up the logging, based on the service name (usually with a `.out` or `.err` suffix). I'll clarify the text and note that `;` is banned. Thinking about it I could also reject `;' in `memlogd` just in case.

Reply:
We should possibly also disallow `;` in service names in the first place, if we don't already